### DISCUSSION TITLE
Debugging an app

### DISCUSSION THREAD

**Discussion by danteppc** [bitcraze/discussions#126] [Q&A] [answered]
Hi,

We are unable to debug the Crazyflie firmware when built with an app. we get the error saying "#include app.h not found". We don't have a problem when debugging the firmware itself.

I find this strange because the Makefile in the firmware already has the INCLUDES and the app's Makefile is set to use all these paths, so what could be the problem here? 

Thank you in advance!

**Comment by knmcguire** [comment]
Hi!

Just some more questions about your issue:
- are you referring to the on-chip debugging, like with these instructions? https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/development/openocd_gdb_debugging/
- How are you building the firmware + app?  are you cd-ing into the app folder and make it there or do you use the toolbelt for this?

Not sure what is going on, but just asking to be able to recreate your problem.

**Reply by danteppc** [reply]
Hi,

Thanks for the quick response,

Yes we followed the instructions as in the page,

We cd to the app directory to "make" and cd to /build to "make flash". (no toolbelt)

Actually just now we were able to debug the app on Eclipse (by pointing to the right cf.elf path) but in VSCode we get the error I said earlier.

Please see the VS launch.json file in the attachment

[launch.txt](https://github.com/bitcraze/discussions/files/9046856/launch.txt)

**Reply by knmcguire** [reply]
Hi! Sorry we didn't get around to test this out this week due to an conference. I've made an todo for ourselves, so I'll hope to be able to recreate your issue.

**Reply by danteppc** [reply]
Thank you, I'd like to add more context maybe this will further clarify our situation: 
1. Right now we use Eclipse and we're able to catch breakpoints if we manually "make" the app and cd to the app's build directory to "make flash"; otherwise we end up debugging the old "cf2.elf". I was wondering if it's possible to tell Eclipse to also build (compile) & flash before debugging? 
2. We don't see the output of "prints" in the console within Eclipse, so we have to launch CFClient and see the outputs there. Is there a way around this?
3. It seems Eclipse is not talking to the gcc compiler because whenever we have a syntax error in the code we rely on Terminal to see if we missed a semicolon or petty syntax or compile errors which is honestly annoying in 2022.
Thanks

**Reply by evoggy** [reply]
Hi,

1) So if I understand correctly you can debug the app, but the issue is that Eclipse cannot tell the difference between build/flash/debug the app and the main firmware unless you change directory manually?
2) You can use this script to connect to the Crazyflie and get the console printouts in a terminal instead: https://github.com/bitcraze/crazyflie-lib-python/blob/master/examples/console/console.py
3) Unfortunately there's currently no one at the office that uses Eclipse (vacation times..) so there's not much I can say about this.

**Comment by knmcguire** [comment]
@danteppc the eclipse instructions are quite outdated here: https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/development/openocd_gdb_debugging/. Could you try to see if the vscode way works for you?

**Reply by danteppc** [reply]
the problem is when we want to debug an app in VSCode we get the error "#include app.h not found" as we also said in the first post. but debugging the same app using Eclipse works.

**Reply by knmcguire** [reply]
aaah oke sorry that was unclear. Sorry, the original post didn't say you were using vscode. 

Sorry for the delay but it seems that the instructions are outdated for the vscode one, as it seems that the latest gdb version doesn't seem to work for cortex debug. Could you let us know which version of openocd and gdb you are using?

```
openocd --version
arm-none-eabi-gdb --version
```

**Reply by danteppc** [reply]
macOS
Openocd version is 0.11.0
GNU gdb (GNU Arm Embedded Toolchain 9-2020-q2-update) 8.3.1.20191211-git
J-Link Debug

please note that we don't have a problem debugging (setting breakpoints tbh) the firmware itself, the problem is only when debugging an app e.g. hello world

**Comment by knmcguire** [comment]
Another thing that might has gone wrong: Have you compiled the firmware with `CONFIG_APP_ENABLE=y` or selected it in the menuconfig > app layer configuration? 

It might be that this configuration goes wrong when compiling and that the app.h is not included but looking at the firmware this should not be a problem.

**Comment by danteppc** [comment, marked as answer]
Ok so thank you for the replies, the problem has been solved. We were trying to run on single file rather than using the proper STM32 debug button 😅

![vs](https://user-images.githubusercontent.com/15725082/179528766-314b0b3f-5084-4702-a81a-b988d709c3f9.jpg)



Sorry for the trouble.

**Reply by knmcguire** [reply]
Ahhh oke, quite a dangerous GUI design 😉 thanks for sharing!

Btw, for our reference, how does your launch file look like for debugging an app? I want to update the instructions with that too, as that is not clear at all

**Reply by danteppc** [reply]
Yeah sure, this is our Launch.json:


```
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        {
            "name": "STM32 Debug",
            "cwd":  "${workspaceRoot}",
            "executable": "${workspaceRoot}/examples/app_hello_world-cpp/build/cf2.elf",
            "request": "launch",
            "type": "cortex-debug",
            "device": "STM32F405",
            "svdFile": "STM32F405.svd",
            "servertype": "openocd",
            "configFiles": ["interface/jlink.cfg", "target/stm32f4x.cfg"],
            "runToMain": true,
            "preLaunchCommands": [
                "set mem inaccessible-by-default off",
                "enable breakpoint",
                "monitor reset"
            ]
        },
        // {
        //     "name": "STM32 App Debug"
        // }
    ]
}
```

By the way, since we're using J-Link debugger, we also had to add "transport select swd" to the interface/jlink.cfg.

**Reply by knmcguire** [reply]
ah thanks a lot!
